File: MtxVec.Delphi/BenchmarkFramework.pas
unit BenchmarkFramework;

interface
{$I bdsppdefs.inc}
uses
  Classes,
  Math387,
  AbstractMtxVec,
  MtxVec,
  BenchmarkResults;            

type
  TVecMethod = function (const src: TMtxVec): TMtxVec of object;
  TSmplFunc  = function (const x: double): double;
  TCplxFunc  = function (const x: TCplx): TCplx;

type
  TFuncInfo = record
    func_name: string;
    smpl_func: TSmplFunc;
    cplx_func: TCplxFunc;
    vec_method: TVecMethod;
    smpl_val: TSample;
    cplx_val: TCplx;
  end;

type
  TCalcProc = procedure of object;
  TBenchmarkFramework = class
  private
    fSrc, fDst: TVec;
    fSrca, fDsta: TSampleArray;
    fSrcac, fDstac: TCplxArray;
    fIterationCount: integer;
    fVectorLength: integer;

    fFuncs: array of TFuncInfo;
    fFuncCount: integer;
    fSelectedFunc: TFuncInfo;

    procedure InitFuncs;
    function GetFuncCount: integer;
    function GetFuncName(idx: integer): string;
    function FindFuncByName (const func_name: string): integer;
    procedure InitSmplVecs (const val: TSample);
    procedure InitCplxVecs (const cval: TCplx);

    procedure CalcVector;
    procedure CalcSample;
    procedure CalcComplex;

    function DoUnderTimer (calc_proc: TCalcProc): double;

  public
    constructor Create;
    destructor Destroy; override;

    function RecalcIterationCount: integer;

    procedure Run (const func_name: string; var results: TBenchmarkResults);

    property VectorLength: integer read fVectorLength write fVectorLength;
    property IterationCount: integer read fIterationCount write fIterationCount;

    property FuncCount: integer read GetFuncCount;
    property FuncName[idx: integer]: string read GetFuncName;
  end;

implementation
uses
  Windows,
  SysUtils
  {$IFDEF CLR}
  ,Math
  {$ENDIF}
  ;

{$IFDEF CLR}
type
  TFuncHelper = class helper for TMtxVec
  private
    function Sin1 (src: TMtxVec): TMtxVec;
    function Cos1 (src: TMtxVec): TMtxVec;
    function Exp1 (src: TMtxVec): TMtxVec;
    function Ln1 (src: TMtxVec): TMtxVec;

    function Log101 (src: TMtxVec): TMtxVec;
    function Tan1 (src: TMtxVec): TMtxVec;
    function ArcTan1 (src: TMtxVec): TMtxVec;
    function ArcSin1 (src: TMtxVec): TMtxVec;

    function ArcCos1 (src: TMtxVec): TMtxVec;
    function ArcTanh1 (src: TMtxVec): TMtxVec;
    function ArcSinh1 (src: TMtxVec): TMtxVec;

    function Sinh1 (src: TMtxVec): TMtxVec;
    function Tanh1 (src: TMtxVec): TMtxVec;
    function Cosh1 (src: TMtxVec): TMtxVec;
  end;

{ TFuncHelper }

function TFuncHelper.Sin1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited Sin (src);
end;

function TFuncHelper.Cos1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited Cos (src);
end;

function TFuncHelper.Exp1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited Exp (src);
end;

function TFuncHelper.Ln1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited Ln (src);
end;

function TFuncHelper.Log101(src: TMtxVec): TMtxVec;
begin
  Result:= inherited Log10 (src);
end;

function TFuncHelper.Tan1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited Tan (src);
end;

function TFuncHelper.ArcTan1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited ArcTan (src);
end;

function TFuncHelper.ArcSin1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited ArcSin (src);
end;

function TFuncHelper.ArcCos1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited ArcCos (src);
end;

function TFuncHelper.ArcTanh1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited ArcTanh (src);
end;

function TFuncHelper.ArcSinh1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited ArcSinh (src);
end;

function TFuncHelper.Tanh1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited Tanh (src);
end;

function TFuncHelper.Sinh1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited Sinh (src);
end;

function TFuncHelper.Cosh1(src: TMtxVec): TMtxVec;
begin
  Result:= inherited Cosh (src);
end;

{$ENDIF}

{ TBenchmarksFramework }

constructor TBenchmarkFramework.Create;
begin
  inherited;
  fSrc := TVec.Create;
  fDst := TVec.Create;
  InitFuncs;
end;

destructor TBenchmarkFramework.Destroy;
begin
  fSrc.Free;
  fDst.Free;
  inherited;
end;

function TBenchmarkFramework.FindFuncByName (const func_name: string): integer;
begin
  for Result:= 0 to fFuncCount-1 do
    if SameText (fFuncs[Result].func_name, func_name) then
      Exit;

  raise Exception.CreateFmt ('TBenchmarkFramework.FindFuncByName: Name of function was not found', [func_name]);
end;

//type TMathFunction = function (a: double): double;

function TBenchmarkFramework.GetFuncCount: integer;
begin
  Result:= fFuncCount;
end;

function TBenchmarkFramework.GetFuncName(idx: integer): string;
begin
  Assert (idx < fFuncCount, Format('TBenchmarkFramework.GetFuncName: Index out of range: %d', [idx]));
  Result:= fFuncs[idx].func_name;
end;

{$IFDEF CLR} //needed because of compiler error when linking directly to System.Math.Sin

function Sin(a: double): double;
begin
    REsult := System.Math.Sin(a);
end;

function Cos(a: double): double;
begin
    REsult := System.Math.Cos(a);
end;

function Exp(a: double): double;
begin
    REsult := System.Math.Exp(a);
end;

function Ln(a: double): double;
begin
    Result := Sin(a);
end;

{$ENDIF}


procedure TBenchmarkFramework.InitFuncs;
const
  VAL1: TSample = 0.5;
  CVAL1: TCplx  = (Re:0.5; Im:0.5);


  procedure Add (const func_name: string; smpl_val: TSample; const cplx_val: TCplx;
                 smpl_func: TSmplFunc; cplx_func: TCplxFunc; vec_method: TVecMethod); overload;
  begin
    if fFuncCount >= Length(fFuncs) then
      SetLength (fFuncs, fFuncCount + 20);

    fFuncs[fFuncCount].func_name:= func_name;
    fFuncs[fFuncCount].smpl_func:= smpl_func;
    fFuncs[fFuncCount].cplx_func:= cplx_func;
    fFuncs[fFuncCount].vec_method:= vec_method;
    fFuncs[fFuncCount].smpl_val:= smpl_val;
    fFuncs[fFuncCount].cplx_val:= cplx_val;

    Inc (fFuncCount);
  end;

begin
  fFuncCount:= 0;

  {$IFNDEF CLR}

    {$IFNDEF CPUX64}

        Add ('Sin', VAL1, CVAL1, Math387.Sin, Math387.Sin, fDst.Sin );
        Add ('Cos', VAL1, CVAL1, Math387.Cos, Math387.Cos, fDst.Cos );
        Add ('Exp', VAL1, CVAL1, Math387.Exp, Math387.Exp, fDst.Exp );
        Add ('Ln', VAL1, CVAL1,  Math387.Ln, Math387.Ln, fDst.Ln );

    {$ELSE}

        {$IFDEF TTDOUBLE}  //define this, if sizeof(TSample) = 8, MtxVec runs in double precision

        Add ('Sin', VAL1, CVAL1, {$IFDEF D23} System.{$ENDIF}Sin, Math387.Sin, fDst.Sin );
        Add ('Cos', VAL1, CVAL1, {$IFDEF D23} System.{$ENDIF}Cos, Math387.Cos, fDst.Cos );
        Add ('Exp', VAL1, CVAL1, {$IFDEF D23} System.{$ENDIF}Exp, Math387.Exp, fDst.Exp );
        Add ('Ln', VAL1, CVAL1,  {$IFDEF D23} System.{$ENDIF}Ln, Math387.Ln, fDst.Ln );

        {$ELSE}

        //There is no single precision of System.Sin, System.Cos, etc... for 64bit

//        Add ('Sin', VAL1, CVAL1, Sin, Math387.Sin, fDst.Sin);
//        Add ('Cos', VAL1, CVAL1, Cos, Math387.Cos, fDst.Cos);
//        Add ('Exp', VAL1, CVAL1, Exp, Math387.Exp, fDst.Exp);
//        Add ('Ln', VAL1, CVAL1,  Ln, Math387.Ln, fDst.Ln);

        {$ENDIF}

    {$ENDIF}

  Add ('Log10', VAL1, CVAL1, Math387.Log10, Math387.Log10, fDst.Log10);
  Add ('Tan', VAL1, CVAL1, Math387.Tan, Math387.Tan, fDst.Tan);

  Add ('ArcSin', VAL1, CVAL1, Math387.ArcSin, Math387.ArcSin, fDst.ArcSin);
  Add ('ArcCos', VAL1, CVAL1, Math387.ArcCos, Math387.ArcCos, fDst.ArcCos);

  {$IFDEF CPUX64}
  Add( 'ArcTan', VAL1, CVAL1, {$IFDEF D23}System.{$ENDIF}ArcTan, Math387.ArcTan, fDst.ArcTan);
  {$ELSE}
  Add( 'ArcTan', VAL1, CVAL1, {$IFDEF D23}Math387.{$ENDIF}ArcTan, Math387.ArcTan, fDst.ArcTan);
  {$ENDIF}

  Add ('Sinh', VAL1, CVAL1, Math387.Sinh, Math387.Sinh, fDst.Sinh);
  Add ('Cosh', VAL1, CVAL1, Math387.Cosh, Math387.Cosh, fDst.Cosh);
  Add ('Tanh', VAL1, CVAL1, Math387.Tanh, Math387.Tanh, fDst.Tanh);

  Add ('ArcSinh', VAL1, CVAL1, Math387.ArcSinh, Math387.ArcSinh, fDst.ArcSinh);
  Add ('ArcCosh', VAL1, CVAL1, Math387.ArcCosh, Math387.ArcCosh, fDst.ArcCosh);
  Add ('ArcTanh', VAL1, CVAL1, Math387.ArcTanh, Math387.ArcTanh, fDst.ArcTanh);

  {$ELSE} //CLR

  Add ('Sin', VAL1, CVAL1, Sin, Math387.Sin, fDst.Sin);
  Add ('Cos', VAL1, CVAL1, Cos, Math387.Cos, fDst.Cos);
  Add ('Exp', VAL1, CVAL1, Exp, Math387.Exp, fDst.Exp);
  Add ('Ln', VAL1, CVAL1, Ln, Math387.Ln, fDst.Ln);

  Add ('Log10', VAL1, CVAL1, Math387.Log10, Math387.Log10, fDst.Log101);
  Add ('Tan', VAL1, CVAL1, Math387.Tan, Math387.Tan, fDst.Tan1);
//  Add ('ArcTan', VAL1, CVAL1, Math387.ArcTan, Math387.ArcTan, fDst.ArcTan1);
  Add ('ArcSin', VAL1, CVAL1, Math387.ArcSin, Math387.ArcSin, fDst.ArcSin1);
  Add ('ArcCos', VAL1, CVAL1, Math387.ArcCos, Math387.ArcCos, fDst.ArcCos1);

  Add ('ArcTanh', VAL1, CVAL1, Math387.ArcTanh, Math387.ArcTanh, fDst.ArcTanh1);
  Add ('ArcSinh', VAL1, CVAL1, Math387.ArcSinh, Math387.ArcSinh, fDst.ArcSinh1);
//  Add ('ArcCosh', VAL1, CVAL1, Math387.ArcCosh, Math387.ArcCosh, fDst.ArcCosh);

  Add ('Tanh', VAL1, CVAL1, Math387.Tanh, Math387.Tanh, fDst.Tanh1);
  Add ('Sinh', VAL1, CVAL1, Math387.Sinh, Math387.Sinh, fDst.Sinh1);
  Add ('Cosh', VAL1, CVAL1, Math387.Cosh, Math387.Cosh, fDst.Cosh1);

  {$ENDIF}
end;

procedure TBenchmarkFramework.InitSmplVecs (const val: TSample);
begin
  fSrc.Size (VectorLength, FALSE);
  fSrc.SetVal (val);
  FSrc.CopyToArray(FSrca);
  fDst.Size (VectorLength, FALSE);
  fDst.SetZero;
  FDst.SizeToArray(FDsta);
end;

procedure TBenchmarkFramework.InitCplxVecs (const cval: TCplx);
begin
  fSrc.Size (VectorLength, TRUE);
  fSrc.SetVal (cval);
  FSrc.CopyToArray(FSrcac);
  fDst.Size (VectorLength, TRUE);
  fDst.SetZero;
  FDst.SizeToArray(FDstac);
end;

procedure TBenchmarkFramework.Run(const func_name: string; var results: TBenchmarkResults);
var
  smpl_vec_ticks, cplx_vec_ticks, smpl_func_ticks, cplx_func_ticks: double;
begin
  fSelectedFunc:= fFuncs[FindFuncByName(func_name)];

  InitSmplVecs (fSelectedFunc.smpl_val);
  smpl_vec_ticks:= DoUnderTimer (CalcVector);

  InitCplxVecs (fSelectedFunc.cplx_val);
  cplx_vec_ticks:= DoUnderTimer (CalcVector);

  InitSmplVecs (fSelectedFunc.smpl_val);
  smpl_func_ticks:= DoUnderTimer (CalcSample);

  InitCplxVecs (fSelectedFunc.cplx_val);
  cplx_func_ticks:= DoUnderTimer (CalcComplex);

  results.Add (func_name, VectorLength, IterationCount,
    smpl_vec_ticks, cplx_vec_ticks, smpl_func_ticks, cplx_func_ticks);
end;

procedure TBenchmarkFramework.CalcVector;
var
  i: integer;
  vec_method: TVecMethod;
begin
  vec_method:= fSelectedFunc.vec_method;
  for i:= 1 to fIterationCount do
    vec_method(fSrc);
end;

procedure TBenchmarkFramework.CalcSample;
var
  i,j: integer;
  smpl_func: TDoubleFunc;
begin
  smpl_func:= fSelectedFunc.smpl_func;
  for i:= 1 to fIterationCount do
    for j:= 0 to fDst.Length-1 do
      fDsta[j]:= smpl_func (fSrca[j]);
end;

procedure TBenchmarkFramework.CalcComplex;
var
  i,j: integer;
  cplx_func: TCplxFunc;
begin
  cplx_func:= fSelectedFunc.cplx_func;
  for i:= 1 to fIterationCount do
    for j:= 0 to fDst.Length-1 do
      fDstac[j]:= cplx_func (fSrcac[j]);
end;

function TBenchmarkFramework.DoUnderTimer(calc_proc: TCalcProc): double;
begin
  StartTimer;
  calc_proc;
  Result:= StopTimer*1000;
end;

function TBenchmarkFramework.RecalcIterationCount: integer;
var
  ticks: DWORD;
  delta: integer;
  a,b: TVec;
  n: integer;
begin
  CreateIt (a,b);
  try
    a.Size (VectorLength, FALSE);
    b.Size (VectorLength, FALSE);

    n:= 0;
//    delta:= 0;
    ticks:= GetTickCount;
    while (TRUE) do begin
      b.Sin(a);
      Inc(n);
      delta:= GetTickCount - ticks;
      if delta > 1000 then
        break;
    end;

    Result:= n div 4;  // iterations per 1/4 second

  finally
    FreeIt (a,b)
  end;
end;

end.

File: MtxVec.CPP/BenchmarkFramework.cpp
//---------------------------------------------------------------------------
#pragma hdrstop

#include "BenchmarkFramework.h"
#include "math.h"
#include "MtxExpr.hpp"
//---------------------------------------------------------------------------

#pragma package(smart_init)

//---------------------------------------------------------------------------
// The Math387 enhanced scalar routines are stored and timed DIRECTLY as
// __fastcall function pointers (TFastDbleFunc) - exactly as the complex column
// already stores Math387 via TCplxFunc. No __cdecl adapter wrapper is used: a
// wrapper would either inline (silently assuming a zero-cost call) or add a
// second call the pure-C/C++ runtime column never pays, distorting the
// comparison. The runtime (<math.h>) versions are timed separately as the
// pure-C baseline (smpl_rtl_func), each likewise stored as a direct pointer.
//
// asinh/acosh/atanh are C99 inverse hyperbolics used by the pure-C/C++ runtime
// baseline column. Modern Embarcadero clang (bcc64x / bcc32x, clang >= 15)
// declares them in <math.h> and links them, so there we time the genuine
// runtime function. Classic bcc32 does not declare them, and the early clang
// Win64 RTL (XE6/D20) declares but does not export them; on those toolchains
// fall back to the standard pure-C log/sqrt identities so the baseline column
// still measures a real scalar computation instead of dropping the demo.
#if defined(__clang__) && (__clang_major__ >= 15)
  #define rt_asinh asinh
  #define rt_acosh acosh
  #define rt_atanh atanh
#else
  static double rt_asinh(double x) { return log(x + sqrt(x * x + 1.0)); }
  static double rt_acosh(double x) { return log(x + sqrt(x * x - 1.0)); }
  static double rt_atanh(double x) { return 0.5 * log((1.0 + x) / (1.0 - x)); }
#endif

//---------------------------------------------------------------------------

/* TBenchmarksFramework */

BenchmarkFramework::BenchmarkFramework()
{
  fSrc = new TVec();
  fDst = new TVec();
//  CreateIt (fSrc, fDst);
  InitFuncs();
}
//---------------------------------------------------------------------------

BenchmarkFramework::~BenchmarkFramework()
{
  delete fSrc;
  delete fDst;	
//  FreeIt (fSrc, fDst);
}
//---------------------------------------------------------------------------

int BenchmarkFramework::FindFuncByName (const AnsiString& func_name)
{
  for (unsigned i = 0; i < fFuncCount; i++)
    if (SameText (fFuncs[i].func_name, func_name))
      return i;

  throw Exception ("BenchmarkFramework::FindFuncByName: Name of function was not found", ARRAYOFCONST((func_name)));
}
//---------------------------------------------------------------------------

int BenchmarkFramework::GetFuncCount()
{
  return fFuncCount;
}
//---------------------------------------------------------------------------

AnsiString BenchmarkFramework::GetFuncName(int idx)
{
  return fFuncs[idx].func_name;
}
//---------------------------------------------------------------------------

void BenchmarkFramework::AddFunc (const AnsiString& func_name, double smpl_val, const TCplx& cplx_val,
								  TFastDbleFunc const smpl_func, TDbleFunc const smpl_rtl_func,
								  TCplxFunc const cplx_func, TVecMethod const vec_method)
{
  if (fFuncCount >= fFuncs.size())
    fFuncs.resize (fFuncCount + 20);

  fFuncs[fFuncCount].func_name = func_name;
  fFuncs[fFuncCount].smpl_func = smpl_func;
  fFuncs[fFuncCount].smpl_rtl_func = smpl_rtl_func;
  fFuncs[fFuncCount].cplx_func = cplx_func;
  fFuncs[fFuncCount].vec_method = vec_method;
  fFuncs[fFuncCount].smpl_val = smpl_val;
  fFuncs[fFuncCount].cplx_val = cplx_val;

  fFuncCount++;
}
//---------------------------------------------------------------------------

void BenchmarkFramework::InitFuncs()
{
  const double VAL1(0.5);
  const TCplx CVAL1(Cplx(0.5,0.5));

  fFuncCount = 0;

  // Per function:  name, real val, cplx val,
  //                Math387 scalar (enhanced),  C/C++ runtime scalar (baseline),
  //                Math387 complex scalar,     MtxVec vectorized method.
  // Order matches BenchmarkFramework.pas so the charts line up across platforms.
  AddFunc ("Sin",     VAL1, CVAL1, Math387::Sin,     sin,   Math387::Sin,     fDst->Sin);
  AddFunc ("Cos",     VAL1, CVAL1, Math387::Cos,     cos,   Math387::Cos,     fDst->Cos);
  AddFunc ("Exp",     VAL1, CVAL1, Math387::Exp,     exp,   Math387::Exp,     fDst->Exp);
  AddFunc ("Ln",      VAL1, CVAL1, Math387::Ln,      log,   Math387::Ln,      fDst->Ln);

  AddFunc ("Log10",   VAL1, CVAL1, Math387::Log10,   log10, Math387::Log10,   fDst->Log10);
  AddFunc ("Tan",     VAL1, CVAL1, Math387::Tan,     tan,   Math387::Tan,     fDst->Tan);
  AddFunc ("ArcSin",  VAL1, CVAL1, Math387::ArcSin,  asin,  Math387::ArcSin,  fDst->ArcSin);
  AddFunc ("ArcCos",  VAL1, CVAL1, Math387::ArcCos,  acos,  Math387::ArcCos,  fDst->ArcCos);
  AddFunc ("ArcTan",  VAL1, CVAL1, Math387::ArcTan,  atan,  Math387::ArcTan,  fDst->ArcTan);

  AddFunc ("Sinh",    VAL1, CVAL1, Math387::Sinh,    sinh,  Math387::Sinh,    fDst->Sinh);
  AddFunc ("Cosh",    VAL1, CVAL1, Math387::Cosh,    cosh,  Math387::Cosh,    fDst->Cosh);
  AddFunc ("Tanh",    VAL1, CVAL1, Math387::Tanh,    tanh,  Math387::Tanh,    fDst->Tanh);

  AddFunc ("ArcSinh", VAL1, CVAL1, Math387::ArcSinh, rt_asinh, Math387::ArcSinh, fDst->ArcSinh);
  AddFunc ("ArcCosh", VAL1, CVAL1, Math387::ArcCosh, rt_acosh, Math387::ArcCosh, fDst->ArcCosh);
  AddFunc ("ArcTanh", VAL1, CVAL1, Math387::ArcTanh, rt_atanh, Math387::ArcTanh, fDst->ArcTanh);
}
//---------------------------------------------------------------------------

void BenchmarkFramework::InitSmplVecs (double val)
{
  fSrc->Size (VectorLength, false);
  fSrc->SetVal (val);
  fSrca.resize(VectorLength);
  for (int i = 0; i < VectorLength; i++)
    fSrca[i] = val;

  fDst->Size (VectorLength, false);
  fDst->SetZero();
  fDsta.resize(VectorLength);
  for (int i = 0; i < VectorLength; i++)
    fDsta[i] = 0;
}
//---------------------------------------------------------------------------

void BenchmarkFramework::InitCplxVecs (const TCplx& cval)
{
  fSrc->Size (VectorLength, true);
  fSrc->SetVal (cval);
  fSrcac.resize(VectorLength);
  for (int i = 0; i < VectorLength; i++)
    fSrcac[i] = cval;

  fDst->Size (VectorLength, true);
  fDst->SetZero();
  fDstac.resize(VectorLength);
  for (int i = 0; i < VectorLength; i++)
    fDstac[i] = C_ZERO;
}
//---------------------------------------------------------------------------

void BenchmarkFramework::Run(const AnsiString& func_name, TBenchmarkResults* results)
{
  fSelectedFunc = fFuncs[FindFuncByName(func_name)];

  InitSmplVecs (fSelectedFunc.smpl_val);
  double smpl_vec_ticks = DoUnderTimer (CalcVector);

  InitCplxVecs (fSelectedFunc.cplx_val);
  double cplx_vec_ticks = DoUnderTimer (CalcVector);

  InitSmplVecs (fSelectedFunc.smpl_val);
  double smpl_func_ticks = DoUnderTimer (CalcSample);       // Math387 enhanced scalar

  InitSmplVecs (fSelectedFunc.smpl_val);
  double smpl_rtl_func_ticks = DoUnderTimer (CalcSampleRtl); // pure C/C++ runtime scalar

  InitCplxVecs (fSelectedFunc.cplx_val);
  double cplx_func_ticks = DoUnderTimer (CalcComplex);

  results->Add (func_name, VectorLength, IterationCount,
    smpl_vec_ticks, cplx_vec_ticks, smpl_func_ticks, smpl_rtl_func_ticks, cplx_func_ticks);
}
//---------------------------------------------------------------------------

void __fastcall BenchmarkFramework::CalcVector()
{
  TVecMethod vec_method = fSelectedFunc.vec_method;
  for (int i = 0; i < fIterationCount; i++)
    vec_method(fSrc);
}
//---------------------------------------------------------------------------

void __fastcall BenchmarkFramework::CalcSample()
{
  TFastDbleFunc smpl_func = fSelectedFunc.smpl_func;
  for (int i = 0; i < fIterationCount; i++)
    for (int j = 0; j < VectorLength; j++)
      fDsta[j] = smpl_func (fSrca[j]);
}
//---------------------------------------------------------------------------

void __fastcall BenchmarkFramework::CalcSampleRtl()
{
  TDbleFunc smpl_func = fSelectedFunc.smpl_rtl_func;
  for (int i = 0; i < fIterationCount; i++)
    for (int j = 0; j < VectorLength; j++)
      fDsta[j] = smpl_func (fSrca[j]);
}
//---------------------------------------------------------------------------

void __fastcall BenchmarkFramework::CalcComplex()
{
  TCplxFunc cplx_func = fSelectedFunc.cplx_func;
  for (int i = 0; i < fIterationCount; i++)
    for (int j = 0; j < VectorLength; j++)
      fDstac[j] = cplx_func (fSrcac[j]);
}
//---------------------------------------------------------------------------

double BenchmarkFramework::DoUnderTimer(TCalcProc calc_proc)
{
  StartTimer();
  calc_proc();
  return StopTimer()*1000;
}
//---------------------------------------------------------------------------

int BenchmarkFramework::RecalcIterationCount()
{
  sVector a,b;
  a.Size (VectorLength, false);
  b.Size (VectorLength, false);

  int n = 0;
  double delta;
  StartTimer();
  while (true) {
	b.Sin(a);
	n++;
    delta = StopTimer()*1000;
    if (delta > 1000)
      break;
  }

  return n / 4; // iterations per 1/4 second
}
//---------------------------------------------------------------------------

